src/toxcore/dht/server/hole_punching.rs

```diff
@@ -0,0 +1,363 @@
+            server.send_ping_req(
+                &PackedNode::new(false, SocketAddr::new(ip, first_port), &friend_pk)
+            )
+        } else {
+            let res = self.first_hole_punching(ports, ip, server, friend_pk);
+
+            self.first_punching_index += MAX_PORTS_TO_PUNCH;
+
+            res
+        };
+
+        let last_hole_punching = if self.num_punch_tries > MAX_NORMAL_PUNCHING_TRIES {
+            let res = self.last_hole_punching(ip, server, friend_pk);
+            self.last_punching_index += MAX_PORTS_TO_PUNCH - (MAX_PORTS_TO_PUNCH / 2);
+
+            res
+        } else {
+            Box::new(future::ok(()))
+        };
+
+        self.num_punch_tries += 1;
+
+        Box::new(first_hole_punching.join(last_hole_punching)
+            .map(|_| ())
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use toxcore::dht::packet::*;
+    use toxcore::binary_io::*;
+    use futures::sync::mpsc;
+    use std::ops::DerefMut;
+    use std::thread;
+
+    #[test]
+    fn hole_punch_new_test() {
+        let hole_punch = HolePunching::new();
+
+        assert!(hole_punch.ping_id != 0);
+    }
+
+    #[test]
+    fn hole_punch_get_major_ip_test() {
+        let (pk, sk) = gen_keypair();
+        let (friend_pk, _friend_sk) = gen_keypair();
+        let (tx, _rx) = mpsc::unbounded::<(DhtPacket, SocketAddr)>();
+        let alice = Server::new(tx, pk, sk);
+        let addrs = vec![];
+
+        let mut hole_punch = HolePunching::new();
+        hole_punch.is_punching_done = false;
+        thread::sleep(Duration::from_millis(1));
+
+        assert!(hole_punch.try_nat_punch(&alice, friend_pk, addrs, Duration::from_millis(1)).wait().is_ok());
+
+        let addrs = vec![
+            "127.0.0.1:11111".parse().unwrap(),
+            "127.0.0.1:22222".parse().unwrap(),
+            "127.0.0.2:33333".parse().unwrap(),
+        ];
+
+        let mut hole_punch = HolePunching::new();
+        hole_punch.is_punching_done = false;
+        thread::sleep(Duration::from_millis(1));
+
+        assert!(hole_punch.try_nat_punch(&alice, friend_pk, addrs, Duration::from_millis(1)).wait().is_ok());
+
+        let addrs = vec![
+            "127.0.0.1:11111".parse().unwrap(),
+            "127.0.0.1:22222".parse().unwrap(),
+            "127.0.0.2:33333".parse().unwrap(),
+            "127.0.0.1:44444".parse().unwrap(),
+            "127.0.0.1:55555".parse().unwrap(),
+            "127.0.0.1:55556".parse().unwrap(),
+            "127.0.0.1:55557".parse().unwrap(),
+            "127.0.0.1:55558".parse().unwrap(),
+            "127.0.0.2:55559".parse().unwrap(),
+        ];
+
+        let mut hole_punch = HolePunching::new();
+        hole_punch.is_punching_done = false;
+        thread::sleep(Duration::from_millis(1));
+
+        assert!(hole_punch.try_nat_punch(&alice, friend_pk, addrs, Duration::from_millis(1)).wait().is_ok());
+    }
+
+    #[test]
+    fn hole_punch_get_nat_ports_test() {
```

Comment:
Please describe what is going on and why you need so many IPs and why

Reply:
A friend can have maximum 8 close node.
If 4 or more close nodes have same IP(with different ports), we consider friend is behind NAT.
So we starts hole punching.